bitswap/network/httpnet/httpnet.go

```
@@ -0,0 +1,477 @@
+	for i, v := range receivers {
+		connectionListeners[i] = v
+	}
+	ht.connEvtMgr = network.NewConnectEventManager(connectionListeners...)
+
+	ht.connEvtMgr.Start()
+}
+
+// Stop stops the connect event manager associated with this network.
+// Other methods should no longer be used after calling Stop().
+func (ht *httpnet) Stop() {
+	ht.connEvtMgr.Stop()
+}
+
+// Ping triggers a ping to the given peer and returns the latency.
+func (ht *httpnet) Ping(ctx context.Context, p peer.ID) ping.Result {
+	return ht.pinger.ping(ctx, p)
+
+}
+
+// Latency returns the EWMA latency for the given peer.
+func (ht *httpnet) Latency(p peer.ID) time.Duration {
+	return ht.pinger.latency(p)
+}
+
+func (ht *httpnet) setCooldown(u *url.URL, t time.Time) {
+	ht.cooldownURLsLock.Lock()
+	ht.cooldownURLs[u.String()] = t
+	ht.cooldownURLsLock.Unlock()
+}
+
+func (ht *httpnet) removeCooldown(u *url.URL) {
+	ht.cooldownURLsLock.Lock()
+	delete(ht.cooldownURLs, u.String())
+	ht.cooldownURLsLock.Unlock()
+}
+
+func (ht *httpnet) senderURLs(p peer.ID) []*senderURL {
+	pi := ht.host.Peerstore().PeerInfo(p)
+	urls := network.ExtractURLsFromPeer(pi)
+	if len(urls) == 0 {
+		return nil
+	}
+
+	senderURLs := make([]*senderURL, len(urls))
+	timeNow := time.Now()
+	ht.cooldownURLsLock.RLock()
+	{
+
+		for i, u := range urls {
+			var cooldown time.Time
+			urlstr := u.String()
+			dl, ok := ht.cooldownURLs[urlstr]
+			if ok && timeNow.Before(dl) {
+				cooldown = dl
+			}
+			senderURLs[i] = &senderURL{
+				url:      u,
+				cooldown: cooldown,
+			}
+		}
+	}
+	ht.cooldownURLsLock.RUnlock()
+	return senderURLs
+}
+
+// SendMessage sends the given message to the given peer. It uses
+// NewMessageSender under the hood, with default options.
+func (ht *httpnet) SendMessage(ctx context.Context, p peer.ID, msg bsmsg.BitSwapMessage) error {
+
+	if len(msg.Wantlist()) == 0 {
+		return nil
+	}
+
+	log.Debugf("SendMessage: %s", p)
+
+	// Note: SendMessage seems to only be used to send cancellations.
```

Verify this as it only seems used by the server. Perhaps it is used by the server to send cancellations. Perhaps we should simply return an error here to signal wrong wiring.

---

afaik this is only used in `server.go` in `sendBlocks()`.